sw/device/silicon_creator/lib/drivers/stdusb.c

```diff
@@ -0,0 +1,119 @@
+// Copyright lowRISC contributors (OpenTitan project).
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/silicon_creator/lib/drivers/stdusb.h"
+
+#include "sw/device/lib/base/macros.h"
+#include "sw/device/silicon_creator/lib/drivers/usb.h"
+
+rom_error_t usb_control_setupdata(usb_control_ctx_t *ctx,
+                                  usb_setup_data_t *setup) {
+  rom_error_t result = kErrorOk;
+  uint32_t zero = 0;
+  switch (setup->request) {
+    case kUsbSetupReqGetDescriptor:
+      switch (setup->value >> 8) {
+        case kUsbDescTypeDevice:
+          usb_ep_transfer(0, (void *)ctx->device_desc,
+                          MIN(setup->length, ctx->device_desc->length),
+                          kUsbTransferFlagsIn);
+          break;
+        case kUsbDescTypeConfiguration: {
+          usb_configuration_descriptor_t *cfg =
+              (usb_configuration_descriptor_t *)ctx->config_desc;
+          usb_ep_transfer(0, cfg, MIN(setup->length, cfg->total_length),
+                          kUsbTransferFlagsIn);
+          break;
+        }
+        case kUsbDescTypeString: {
+          size_t index = setup->value & 0xFF;
+          size_t n = 0;
+          while (ctx->string_desc[n] != NULL) {
+            ++n; /* count number of string descriptor */
+          }
+          if (index < n) {
+            usb_ep_transfer(
+                0, (void *)ctx->string_desc[index],
+                MIN(setup->length, (uint8_t)ctx->string_desc[index][0]),
+                kUsbTransferFlagsIn);
+          } else {
+            result = kErrorUsbBadSetup;
+          }
+          break;
+        }
+        default:
+          result = kErrorUsbBadSetup;
+      }
+      break;
+
+    case kUsbSetupReqSetAddress:
+      ctx->next.device_address = (uint8_t)setup->value & 0x7f;
+      usb_ep_transfer(0, NULL, 0, kUsbTransferFlagsIn);
+      ctx->flags |= kUsbControlFlagsPendingAddress;
+      break;
+
+    case kUsbSetupReqSetConfiguration:
+      ctx->next.configuration = (uint8_t)setup->value;
+      usb_ep_transfer(0, NULL, 0, kUsbTransferFlagsIn);
+      ctx->flags |= kUsbControlFlagsPendingConfig;
```

Comment:
_This_ request is supposed to be serviced before the Status stage is completed. I guess if there isn't much happening on the device as a response, perhaps this will amount to "fine" in practice, but if there is considerable setup to activate a configuration, then that must all be completed before responding to the IN transaction with anything but NAK.

Reply:
Fixed: `SetConfiguration` no longer finishes the transaction.  I've updated the documentation to list caller's responsibilities for `SetAddress` and `SetConfiguration`.